lib/streamlit/runtime/context.py

```diff
@@ -0,0 +1,154 @@
+    if not is_type(
+        session_client,
+        "streamlit.web.server.browser_websocket_handler.BrowserWebSocketHandler",
+    ):
+        return None
+    return cast("RequestHandler", session_client).request
+
+
+@lru_cache
+def _normalize_header(name: str) -> str:
+    """Map a header name to Http-Header-Case.
+
+    >>> _normalize_header("coNtent-TYPE")
+    'Content-Type'
+    """
+    return "-".join(w.capitalize() for w in name.split("-"))
+
+
+class StreamlitHeaders(Mapping[str, str]):
+    def __init__(self, headers: Iterable[tuple[str, str]]):
+        dict_like_headers: dict[str, list[str]] = {}
+
+        for key, value in headers:
+            header_value = dict_like_headers.setdefault(_normalize_header(key), [])
+            header_value.append(value)
+
+        self._headers = dict_like_headers
+
+    @classmethod
+    def from_tornado_headers(cls, tornado_headers: HTTPHeaders) -> StreamlitHeaders:
+        return cls(tornado_headers.get_all())
+
+    def get_all(self, key: str) -> list[str]:
+        return list(self._headers.get(_normalize_header(key), []))
+
+    def __getitem__(self, key: str) -> str:
+        try:
+            return self._headers[_normalize_header(key)][0]
+        except LookupError:
+            raise KeyError(key) from None
+
+    def __len__(self) -> int:
+        """Number of unique headers present in request."""
+        return len(self._headers)
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._headers)
+
+    def to_dict(self) -> dict[str, str]:
+        return {key: self[key] for key in self}
+
+
+class StreamlitCookies(Mapping[str, str]):
+    def __init__(self, cookies: Mapping[str, str]):
+        self._cookies = MappingProxyType(cookies)
+
+    @classmethod
+    def from_tornado_cookies(
+        cls, tornado_cookies: dict[str, Morsel[Any]]
+    ) -> StreamlitCookies:
+        dict_like_cookies = {}
+        for key, morsel in tornado_cookies.items():
+            dict_like_cookies[key] = morsel.value
+        return cls(dict_like_cookies)
+
+    def __getitem__(self, key: str) -> str:
+        return self._cookies[key]
+
+    def __len__(self) -> int:
+        """Number of unique headers present in request."""
+        return len(self._cookies)
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._cookies)
+
+    def to_dict(self) -> dict[str, str]:
+        return dict(self._cookies)
+
+
+class ContextProxy:
+    """An interface to context about the current user session. Context is exposed as
+    properties on `st.context`, such as `st.context.headers` or `st.context.cookies`.
+    Each property description explains more about the contents."""
+
+    @property
+    @gather_metrics("context.headers")
+    def headers(self) -> StreamlitHeaders:
+        """A read-only, dict-like access to headers sent in the initial session websocket request. Keys are case-insensitive. Use get_all() to see all values if the same header is set multiple times."""
```

Comment:
nit: docstring line lengths. I think we can probably shorten this summary a bit and provide extra info in the docstring body

Reply:
Fixed!

I added this for better rendering when `st.write(st.context)` is used (the default behavior takes doc 1st line). 

But after discussion with product we decided that is is okay to have a short version in `st.write`, and just normal docstring for developers in code. Changed to multi-line version.